src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingWatcher.cs

```diff
@@ -129,93 +116,43 @@ private FileChangeList ComputeChangesAndUpdateState()
             return changes;
         }
 
-        // returns true for empty, '.', and '..' 
-        private unsafe bool IsSpecial(char* cFileName)
-        {
-            if (cFileName[0] == 0) return true;
-            if (cFileName[0] == '.') {
-                if (cFileName[1] == 0) return true;
-                if (cFileName[1] == '.' && cFileName[2] == 0) return true;
-            }
-            return false;
-        }
-
-        private unsafe static int GetLength(char* nullTerminatedString)
-        {
-            int length = -1;
-            while (true)
-            {
-                if (nullTerminatedString[++length] == 0)
-                {
-                    break;
-                }
-            }
-            return length;
-        }
-        private unsafe static bool EndsWith(char* nullTerminatedString, int length, string possibleEnding)
-        {
-            if(possibleEnding.Length > length)
-            {
-                return false;
-            }
-
-            var start = nullTerminatedString + (length - possibleEnding.Length);
-
-            for(int i=0; i<possibleEnding.Length; i++)
-            {
-                if (start[i] != possibleEnding[i]) return false;
-            }
-            return true;
-        }
-
-        private unsafe bool IsWatched(char* filename)
+        private bool IsWatched(ref FileSystemEntry entry)
         {
+            if (entry.IsDirectory) return false;
             if (_extensionsToWatch == null) return true;
-            var length = GetLength(filename);
-            foreach(var extension in _extensionsToWatch)
+            foreach (var extension in _extensionsToWatch)
             {
-                if(EndsWith(filename, length, extension))
-                {
+                if (Path.GetExtension(entry.FileName).SequenceEqual(extension))
```

Comment:
There is an `EndsWith()` extension for `ReadOnlySpan<char>` that also takes a `StringComparison`. It is in the `System` namespace and comes from the `System.Memory` library (which is part of 2.1).

Reply:
As I mentioned above, it will either be wrong for Windows or wrong for Linux (or we do if defines). My [proposal](https://github.com/dotnet/corefxlab/issues/2222) is that we remove this code and let the user define this part through `ShouldRecursePredicate`.

Reply:
Specifying StringComparison (or `bool ignoreCase`) as an option is the best place to start on this change if you want to expose right away. You can, by default, pick the one you think is applicable based on the platform. MacOS & Windows should be insensitive by default.

I'll review your proposal as soon as I get a chance.